hack/common.sh

```diff
@@ -862,4 +865,101 @@ function os::build::find-binary() {
+# and OS_RELEASE_DOCKER_ARGS can be used to customize the container. The docker socket
+# is mounted by default and the output of the command is the container id.
+os::build::environment::create() {
+  golang_version="${OS_BUILD_ENV_GOLANG}"
+  release_image="${OS_BUILD_ENV_IMAGE}"
+  additional_context="${OS_BUILD_ENV_DOCKER_ARGS:-}"
+  if [[ -z "${additional_context}" && "${OS_BUILD_ENV_USE_DOCKER:-y}" == "y" ]]; then
+    additional_context="--privileged -v /var/run/docker.sock:/var/run/docker.sock"
+
+    if [[ "${OS_BUILD_ENV_LOCAL_DOCKER:-n}" == "y" ]]; then
+      # if OS_BUILD_ENV_LOCAL_DOCKER==y, add the local OS_ROOT as the bind mount to the working dir
+      # and set the running user to the current user
+      workingdir=$( os::build::environment::release::workingdir )
+      additional_context="${additional_context} -v ${OS_ROOT}:${workingdir} -u $(id -u)"
+    elif [[ -n "${OS_BUILD_ENV_REUSE_VOLUME:-}" ]]; then
+      # if OS_BUILD_ENV_REUSE_VOLUME is set, create a docker volume to store the working output so
+      # successive iterations can reuse shared code.
+      workingdir=$( os::build::environment::release::workingdir )
+      name="$( echo "${OS_BUILD_ENV_REUSE_VOLUME}" | tr '[:upper:]' '[:lower:]' )"
+      docker volume create --name "${name}" > /dev/null
+      additional_context="${additional_context} -v ${name}:${workingdir}"
+    fi
+  fi
+
+  # Create a new container to from the release environment
+  docker create ${additional_context} "${release_image}" "$@"
+}
+readonly -f os::build::environment::create
+
+# os::build::environment::release::workingdir calculates the working directory for the current
+# release image.
+os::build::environment::release::workingdir() {
+  # get working directory
+  local container=$(docker create "${release_image}")
+  local workingdir=$(docker inspect -f '{{ index . "Config" "WorkingDir" }}' "${container}")
+  docker rm "${container}" > /dev/null
+  echo "${workingdir}"
+}
+readonly -f os::build::environment::release::workingdir
+
+# os::build::environment::cleanup stops and removes the container named in the argument
+# unless OS_BUILD_ENV_LEAVE_CONTAINER is set to a value besides 'n',
+os::build::environment::cleanup() {
+  container=$1
+  docker stop --time=0 "${container}" > /dev/null || true
+  if [[ "${OS_BUILD_ENV_LEAVE_CONTAINER:-n}" == "n" ]]; then
+    docker rm "${container}" > /dev/null
+  fi
+}
+readonly -f os::build::environment::cleanup
+
+# os::build::environment::withsource starts the container provided as the first argument
+# after copying in the contents of the current Git repository at HEAD (or, if specified,
+# the ref specified in the second argument).
+os::build::environment::withsource() {
+  container=$1
+  commit=${2:-HEAD}
+
+  # Generate version definitions.
+  OS_GIT_TREE_STATE=clean # set this because we will be pulling from git archive
+  os::build::get_version_vars
+
+  if [[ "${OS_BUILD_ENV_LOCAL_DOCKER:-n}" == "y" ]]; then
```

Comment:
This isn't particularly idiomatic... if you want someone to set an envar to trigger something, just use
```sh
if [[ -n "${ENV_TRIGGER:-}" ]]; then
```

If you want someone to have negative flags (which I'd suggest we don't start to introduce) use `-z`. We should strive to use one (idiomatic) paradigm for this sort of thing.